### PR TITLE
Adding Stash connection "IdentifyMetadata" Task

### PR DESCRIPTION
**Pull request by AnonumousMouse**
#### Database Migration
NO

#### Description
Adds option to also schedule a IdentifyMetadata task when sending updates to stash.  There are likely better ways to accomplish this but I am not that familiar with C#/*arr code and just wanted to get something that should work as a starting point. There definitely should be more customization/options in a final version (ex. querying for available scrapers and displaying a selectable list of which to include).  Please feel free to let me know of anything that needs to change or take it over yourself to finalize if there is nothing too major. 

I have successfully tested the query in the graphql playground and it should work although I am not certain about the code changes themselves working. I did not have time to test that and I am hoping the changes are minimal enough that it should work despite my limited familiarity/testing.

#### Todos
- [ ] Tests
- [ ] Translation Keys (./src/NzbDrone.Core/Localization/Core/en.json)
- [ ] [Wiki Updates](https://wiki.servarr.com)

#### Issues Fixed or Closed by this PR

* #65

### CODE REVIEW

File: src/NzbDrone.Core/Notifications/Stash/StashSettings.cs [hunk 1]
```diff
@@ -65,6 +65,9 @@ public StashSettings()
         [FieldDefinition(10, Label = "Map Paths To", Type = FieldType.Textbox, Advanced = true, HelpText = "Stash Path, Used to modify site paths when Stash sees library path location differently from Whisparr")]
         public string MapTo { get; set; }
 
+        [FieldDefinition(11, Label = "Run Identify Task", Type = FieldType.Textbox, Advanced = true, HelpText = "Run metadataIdentify task with StashDB/Autotag on new files")]
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Stash/StashSettings.cs]: Shouldn't this be a `bool` and `FieldType.Checkbox`

File: src/NzbDrone.Core/Notifications/Stash/StashProxy.cs [hunk 2]
```diff
@@ -23,8 +23,22 @@ public void Update(StashSettings settings, string path)
             var generatePreviews = settings.GeneratePreviews ? "true" : "false";
             var cleanPath = path.ToJson();
 
+
+            var MetadataIdentifyQuery = "";
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Stash/StashProxy.cs]: local variables should be camel case (first letter lower)

File: src/NzbDrone.Core/Notifications/Stash/StashProxy.cs [hunk 3]
```diff
@@ -37,6 +51,7 @@ public void Update(StashSettings settings, string path)
                                 scanGeneratePhashes: {(settings.GeneratePhashes ? "true" : "false")},
                                 paths: [{cleanPath}]
                             }})
+                            {(settings.MetadataIdentify ? MetadataIdentifyQuery : "" )}
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Stash/StashProxy.cs]: This is the same logic as you already have above, no use in another conditional here

**Review by Qstick — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/Notifications/Stash/StashProxy.cs [hunk 4]
```diff
@@ -23,8 +23,22 @@ public void Update(StashSettings settings, string path)
...
+                                                {{source: {{scraper_id: ""builtin_autotag""}}, options: {{setOrganized: false}} }}
+                                            ],
+                                            options: {{includeMalePerformers: false, setCoverImage: true, setOrganized: false}}, 
+                                            paths: [{cleanPath}]
+                                       }})";
+            }
+
             request.SetContent(new
-            {
+            {
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Stash/StashProxy.cs]: Also, get rid of this whitespace

File: src/NzbDrone.Core/Notifications/Stash/StashSettings.cs (hunk 1 above)
Comment (AnonumousMouse) [reply src/NzbDrone.Core/Notifications/Stash/StashSettings.cs]: Yep, thanks for catching that.

File: src/NzbDrone.Core/Notifications/Stash/StashProxy.cs [hunk 5]
```diff
@@ -23,6 +23,19 @@ public void Update(StashSettings settings, string path)
             var generatePreviews = settings.GeneratePreviews ? "true" : "false";
             var cleanPath = path.ToJson();
 
+            var metadataIdentifyQuery = "";
+            if (settings.MetadataIdentify) {
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Stash/StashProxy.cs]: compile lint error: `error SA1500: Braces for multi-line statements should not share line`

```
if (settings.MetadataIdentify) 
{
```

File: src/NzbDrone.Core/Notifications/Stash/StashProxy.cs [hunk 6]
```diff
@@ -23,6 +23,19 @@ public void Update(StashSettings settings, string path)
...
+                metadataIdentifyQuery =  $@"metadataIdentify(
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Stash/StashProxy.cs]: Double space

```suggestion
                metadataIdentifyQuery = $@"metadataIdentify(
```

**Review by Qstick — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/Notifications/Stash/StashProxy.cs [hunk 7]
```diff
@@ -23,6 +23,20 @@ public void Update(StashSettings settings, string path)
             var generatePreviews = settings.GeneratePreviews ? "true" : "false";
             var cleanPath = path.ToJson();
 
+            var metadataIdentifyQuery = "";
+            if (settings.MetadataIdentify)
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Stash/StashProxy.cs]:
```suggestion
            if (settings.MetadataIdentify)
```

Comment (AnonumousMouse) [reply src/NzbDrone.Core/Notifications/Stash/StashProxy.cs]: Sorry only just spotted that error in the azure build log

**Comment by Qstick** [comment]
Thanks, got tests good now. Now to investigate the query and options you have. I asked a few people in discord to have a peek.

**Comment by JoeNyland** [comment]
It's cool that you've got a version in progress -- many thanks! Would you consider adding the options that I have in my query and the `updatePerformers` task, too? https://github.com/Whisparr/Whisparr/issues/65#issuecomment-1593620576 I'd have a go myself, but I have zero familiarity with the codebase.

**Comment by AnonumousMouse** [comment]
> It's cool that you've got a version in progress -- many thanks! Would you consider adding the options that I have in my query and the `updatePerformers` task, too? [#65 (comment)](https://github.com/Whisparr/Whisparr/issues/65#issuecomment-1593620576) I'd have a go myself, but I have zero familiarity with the codebase.

Sure, one question I do have is how exactly does that updatePerformers task work? Is it just mass updating all performers from stashdb? Or is it somehow contextually only updating the performers that where just added? If this mutation query was to be added, it would have to be specific in its updates to only elements directly related to the scenes recently changed. Obviously, we can't be calling mass updates on every file grab.

**Comment by JoeNyland** [comment]
> Sure, one question I do have is how exactly does that updatePerformers task work? Is it just mass updating all performers from stashdb?

So the identify task will create a performer for each that it finds in a matched scene. This performer will have a Stash ID, linking to the performer on StashDB, but it won't have any information, other than the performer's name.

The `updatePerformers` query that I've provided will go through and look for performers which have a Stash ID and refresh the information it has for them. If a performer has already been refreshed in a previous run of the task, then it won't be updated again.

> Or is it somehow contextually only updating the performers that where just added?

There's no such information available to be able to know which performers were just added, other than the hint that a performer has a Stash ID.

> If this mutation query was to be added, it would have to be specific in its updates to only elements directly related to the scenes recently changed.

There's just no way of achieving this; the state of 'scenes recently changed' isn't available as each mutation is sort of 'stateless'. There's also no mapping of a path to a performer. So you can't say "I just added /path/to/scene.mp4"; now update the performer(s) for that scene" as there's no way (that I'm aware of) to control the scope of the `updatePerformers` task based on FS paths.

However, as I've explained above, even though a performer who has already been updated previously would be included in this, they won't actually be updated if they already have been updated. So IMO, I don't see an issue in running it.

**Comment by AnonumousMouse** [comment]
> > Sure, one question I do have is how exactly does that updatePerformers task work? Is it just mass updating all performers from stashdb?
> 
> So the identify task will create a performer for each that it finds in a matched scene. This performer will have a Stash ID, linking to the performer on StashDB, but it won't have any information, other than the performer's name.
> 
> The `updatePerformers` query that I've provided will go through and look for performers which have a Stash ID and refresh the information it has for them. If a performer has already been refreshed in a previous run of the task, then it won't be updated again.
> 
> > Or is it somehow contextually only updating the performers that where just added?
> 
> There's no such information available to be able to know which performers were just added, other than the hint that a performer has a Stash ID.
> 
> > If this mutation query was to be added, it would have to be specific in its updates to only elements directly related to the scenes recently changed.
> 
> There's just no way of achieving this; the state of 'scenes recently changed' isn't available as each mutation is sort of 'stateless'. There's also no mapping of a path to a performer. So you can't say "I just added /path/to/scene.mp4"; now update the performer(s) for that scene" as there's no way (that I'm aware of) to control the scope of the `updatePerformers` task based on FS paths.
> 
> However, as I've explained above, even though a performer who has already been updated previously would be included in this, they won't actually be updated if they already have been updated. So IMO, I don't see an issue in running it.

I guess that should probably be fine but I would probably make it a separate toggle though.

**Comment by JoeNyland** [comment]
Agreed, 100%

**Comment by Qstick** [comment]
The feedback from the community on Discord, with `SetOrganized` and `IncludeMalePerformers` being options

```
            variables = {"input": {
              "sources": [
                    {
                    "source":{
                      "stash_box_endpoint":"https://metadataapi.net/graphql"
                        }
                    },
                    {
                    "source": {
                        "scraper_id": "builtin_autotag"
                        },
                        "options": {
                            "setOrganized": False
                          }
                        }
                    ],
              "options":{
                "includeMalePerformers": False,
                "setOrganized": True,
                "setCoverImage": True,
                "fieldOptions":[
                  {
                    "field":"studio",
                    "strategy":"MERGE",
                    "createMissing": True
                  },
                  {
                    "field":"performers",
                    "strategy":"MERGE",
                    "createMissing": True
                  },
                  {
                    "field":"tags",
                    "strategy":"MERGE",
                    "createMissing": True
                  }
                  ]
              },
              "sceneIDs": [],
              "paths": []
              }}
```

**Pull request closed without merging**